implementations/rust/message/src/message.rs

```diff
@@ -98,17 +73,49 @@ pub mod message {
                     return Err(s);
                 }
             }
+
             msg.message_body.append(&mut (w.to_vec()));
             Ok((msg, w))
         }
     }
 
     /* Addresses */
     #[repr(C)]
+    #[derive(Debug, PartialEq)]
+    #[repr(C)]
+    #[derive(Clone, Copy)]
+    pub struct LocalAddress {
+        pub address: u32,
+    }
+
+    #[derive(Clone, Copy)]
+    pub struct RouterAddress {
+        pub a_type: AddressType,
+        pub length: u8,
+        pub address: Address,
+    }
+
+    #[repr(C)]
+    #[derive(Clone, Copy, Debug, PartialEq)]
+    pub enum Address {
+        // First two fields are type and length
+        LocalAddress(LocalAddress),
+        TcpAddress(IpAddr, u16),
+        UdpAddress(IpAddr, u16),
+        ChannelAddress(u32),
+    }
+
+    pub enum HostAddressType {
+        Ipv4 = 0,
+        Ipv6 = 1,
+    }
+
     pub enum AddressType {
+        Undefined = 255,
```

Comment:
Could we change this to use Option instead where Undefined is None and everything else is Some?

Reply:
We could, but then we have to pattern-match every time we want to use AddressType. I'm making the call that it isn't worth it, at least not now.